Feature/pno 12 2

azech-hqs opened the pull request.

azech-hqs committed:
feat: add graph schema with examples

azech-hqs committed:
chore: add husky folder to gitignore

azech-hqs committed:
feat: add parameter schema with examples

azech-hqs committed:
feat: adjust method schema to support unit methods

azech-hqs committed:
feat: add plane wave expansion

azech-hqs committed:
feat: add storage schemas and examples

azech-hqs committed:
feat: add schema for length

azech-hqs committed:
chore: adjust energy schema

Removed requirement for name attribute to be consistent with the length schema.

azech-hqs committed:
feat: add pseudopotential schema and example

azech-hqs committed:
style: use definitions for metadata properties

azech-hqs committed:
refactor: use methodCategoryPaths instead of tiers

azech-hqs committed:
Merge branch 'feature/PNO-7' into feature/PNO-12-2

# Conflicts:
#	example/method.json

timurbazhirov commented on schema/method_units/physical/psp/file/origin.json:
```diff
@@ -0,0 +1,45 @@
+        "ppSources": {
+            "description": "",
+            "enum": [
+                "quantum espresso database",
```
maybe just `espresso` would suffice?

azech-hqs replied:
agreed, these should be slugs anyway

timurbazhirov commented on schema/method_units/physical/psp/file/origin.json:
```diff
@@ -0,0 +1,45 @@
+            "enum": [
+                "quantum espresso database",
+                "gbrv",
+                "vasp database",
```
Same as above - just `vasp` would be sufficient, probably. The pseudopotentials are tied to a specific version of the code also - as long as the above `version` keyword tracks it already, we should be OK.

timurbazhirov commented on schema/core/primitive/acyclic_graph/base_node.json:
```diff
@@ -0,0 +1,42 @@
+        "flowchartId"
+    ],
+    "definitions": {
+        "string-or-array-of-strings": {
```
1. Not sure where this is used
2. If we do need it, maybe put in a definitions file at the `primitive` level?

azech-hqs replied:
This was written to support directed acyclic graphs, such that the `next` field either contains one node ID or a list of node IDs. To my knowledge, this definition isn't currently used. Ok, I will move it to `primitive` then.

timurbazhirov commented on schema/core/primitive/acyclic_graph/base_node.json:
```diff
@@ -0,0 +1,42 @@
+                }
+            ]
+        },
+        "flowchart-id": {
```
1. Should we define the above property on the line 7 through this?
2. Having a "-" in the name may make the schemaIds resolution a bit clunky unless we change the separator in the ID

timurbazhirov commented on schema/core/primitive/acyclic_graph.json:
```diff
@@ -0,0 +1,21 @@
+{
```
For brevity we can drop `acyclic` from the directory name and put in description

timurbazhirov replied:
Would also help with the "-" vs "_" in schema ids

timurbazhirov commented on schema/core/primitive/acyclic_graph/named_node_in_group.json:
```diff
@@ -0,0 +1,28 @@
+            "$ref": "#/definitions/group-info"
+        }
+    ],
+    "definitions": {
```
I think this definition deserves to be in its own file at the top level of the `primitive` folder, no?

azech-hqs replied:
Since we have plans to put it to use in other places, yes. Although generally, I think it is perhaps better to wait to extract definitions like this to separate files until they can be reused in at least one other schema.

timurbazhirov commented on schema/core/reusable/param_int.json:
```diff
@@ -0,0 +1,22 @@
+        "units": {
+            "description": "Unit of parameter",
+            "type": "string"
+        },
```
The two below can be inherited from `scalar`, no?

azech-hqs replied:
I don't think we can use `scalar` in more places than just `param_number.json`. The field `units` and `precision` cannot be inherited from another parameter schema with `allOf`, but we could use definitions/references. I'll include a draft of this in the next version.

timurbazhirov commented on schema/core/reusable/param_int.json:
```diff
@@ -0,0 +1,22 @@
+{
```
maybe let's have a folder `parameters/` and nest integer, number, and string inside it?

azech-hqs replied:
agreed. They are thematically and structurally very similar.

timurbazhirov commented on schema/method_units/mathematical/fapprx.json:
```diff
@@ -15,6 +15,22 @@
+            },
+            "required": [
+                "targetFunction"
+            ]
```
Let's just include `targetFunction` key at the top level instead of putting it in metadata?

timurbazhirov commented on schema/method_units/mathematical/fapprx/basisexp.json:
```diff
@@ -23,6 +23,29 @@
+                        "basisFunctionType"
+                    ]
+                }
+            ]
```
The above implementation seems a bit convoluted, do we need to reuse the whole `metadata` field or just expose the `basisFunctionType` at the top level?

timurbazhirov commented on schema/method_units/physical/psp.json:
```diff
@@ -0,0 +1,60 @@
+                "anyOf": [
+                    {
+                        "enum": [
+                            "linalg/lintra"
```
Wasn't it basis expansion?

azech-hqs replied:
I was referring to projection here. Perhaps this should be then `linalg/lintra/proj`.

timurbazhirov commented on schema/method_units/physical/psp.json:
```diff
@@ -0,0 +1,60 @@
+            "enum": [
+                {
+                    "name": "Pseudopotential Approximation",
+                    "slug": "psp"
```
I remember there was a type definition for oneOf "string" or "object" with name and slug, so that `type: "psp" would also validate, wasn't there?

azech-hqs replied:
yes, there's `slugified_entry_or_slug` in `core/primitive`.

timurbazhirov commented on schema/method_units/physical/psp/file.json:
```diff
@@ -0,0 +1,58 @@
+                "none",
+                "1c-scalar",
+                "2c-quasi",
+                "4c-full"
```
Where did the 1c, 2c, 4c come from? Should we include a link?

timurbazhirov replied:
Speaking about links - should we consider including a DOI for the physical methods like pseudopotential?

azech-hqs replied:
These slugs refer to 1-component, 2-component, and 4-component relativistic Hamiltonians. On this subject, I find this article very helpful:  https://hal.archives-ouvertes.fr/hal-00662643

azech-hqs replied:
Regarding the DOIs, we could include the literature reference schema as part of `metadata` for example.

timurbazhirov replied:
Let's have a task to come back to this

timurbazhirov replied:
And keep the link in description (https://hal.archives-ouvertes.fr/hal-00662643)

azech-hqs committed:
refactor: rename to graph and isolate group info

azech-hqs committed:
refactor: move parameter schemas and isolate units and precision

azech-hqs committed:
chore: simplify ppSources enum

azech-hqs committed:
chore: add schema for string or string array

azech-hqs committed:
chore: adjust tests

azech-hqs committed:
chore: add intermediate base schemas for unit method

Database storage and form population require different logic. The additional schemas separate this logic.

azech-hqs committed:
chore: remove slug from graph schema

timurbazhirov commented on schema/core/primitive/group_info.json:
```diff
@@ -0,0 +1,20 @@
+    "type": "object",
+    "properties": {
+        "groupName": {
+            "description": "Human-readable name of group of node",
```
nodes

timurbazhirov commented on schema/core/primitive/group_info.json:
```diff
@@ -0,0 +1,20 @@
+            "maxLength": 512
+        },
+        "groupId": {
+            "description": "Unique identifier of group node belongs to",
```
the group a node belongs to

timurbazhirov commented on schema/core/reusable/parameters/param_int.json:
```diff
@@ -0,0 +1,19 @@
+{
```
we can just call this and other files in this folder `int`. `string`, `number`

timurbazhirov commented on schema/method_units/physical/psp/set.json:
```diff
@@ -0,0 +1,61 @@
+        "name",
+        "files"
+    ]
+}
```
This will need more clarification. Let's put a task into the epic for the pseudopotential sets / workflow datasets and link to this file/PR to save the work. Then remove the file for now.

timurbazhirov commented on schema/method_units/unit_base_physical.json:
```diff
@@ -0,0 +1,41 @@
+                    "description": "Tracks from which mathematical unit methods this physical method derives",
+                    "type": "array",
+                    "items": {
+                        "items": {
```
Shouldn't be nested

timurbazhirov commented on schema/method_units/unit_base_physical.json:
```diff
@@ -0,0 +1,41 @@
+                        }
+                    }
+                },
+                "modelCategoryPaths": {
```
1. Let's add a path item to reusable and put a note to come back to define it through regex when newer version of JSON schema is adopted
2. Let's define an array of paths and use here

azech-hqs committed:
revert: remove pseudopotential set

azech-hqs committed:
style: rename parameter schemas

azech-hqs committed:
revert: remove storage schemas

azech-hqs committed:
chore: add schemas for path and path array

azech-hqs committed:
revert: remove metadata definitions

azech-hqs committed:
refactor: make composition more readable

azech-hqs committed:
chore: spelling and other fixes

azech-hqs committed:
revert: remove length constraint for string

timurbazhirov committed:
Update unit_base_mathematical.json
